### ISSUE TITLE
Not possible to re-mount if host element is unmounted

### ISSUE THREAD

**Issue by johnwiseheart** [theia-ide/theia#5414]
### Description
I am running theia inside of another element by rebinding `FrontendApplication`.

As an example, instead of binding to the document `body`, I am binding to a `div` inside the body. as I wish for theia to live in a "tab" of a larger application, I wish to arbitrarily hide and re-open it, by unmounting and remounting its host element. 

It would seem like unmounting theia from its host element works fine, but calling `application.start()` presents a variety of issues:

* All commands have already been registered and can't be registered again
* Toolbar items are already registered
* `InMemoryResources` are never removed, so when they are added again we see an error
* etc

This got me thinking - in a perfect world, I don't want to completely remove theia when I swap application tab - just unmount the shell until the user goes back to it. It would be great to keep everything already registered, and in memory so that going back to it, it would load much faster. In this way, my ideal goal would be to add a "reattach" method to the FrontendApplication, which I then tried to do:

```typescript
    class AuthoringFrontendApplication extends FrontendApplication {
        protected getHost(): Promise<HTMLElement> {
            return Promise.resolve(host);
        }

        public attach(el: HTMLElement) {
            // Widget.detach(this.shell);
            this.attachShell(el);
        }
    };
```

However, with this approach, I still don't get what I am looking for, since when I just try to reattach, I get an error that the widget is _already_ attached. However, when I detach first (commented out), I get an error that the widget is already detached.

Any thoughts here?

**OS and Theia version:**

Theia 5.0, but fairly sure this isn't fixed in latest as I think I am the first to be trying to do something like this.

**Comment by akosyakov** [comment]
> It would seem like unmounting theia from its host element works fine, but calling application.start() presents a variety of issues:

In the perfect world calling `ApplicationShell.start` should return `Disposable` object which can be used to unload everything. But it will be quiet a refactoring for all extensions. Having an issue can be good but we don't do it any time soon.

> However, with this approach, I still don't get what I am looking for, since when I just try to reattach, I get an error that the widget is already attached. However, when I detach first (commented out), I get an error that the widget is already detached.

It sounds bogus. Did you manage to investigate why `Widget.detach` throws? I assume that you've already called `start` and the shell was successfully attached.

**Comment by johnwiseheart** [comment]
> In the perfect world calling ApplicationShell.start should return Disposable object which can be used to unload everything. But it will be quiet a refactoring for all extensions. Having an issue can be good but we don't do it any time soon.

Agreed, but understand thats a pretty huge change and not worth prioritising over other things for the moment.

> It sounds bogus. Did you manage to investigate why `Widget.detach` throws? I assume that you've already called `start` and the shell was successfully attached.

Haven't done a whole lot of investigation yet, will do some more today. Do you think just reattaching in this way will "just work"? I don't have a lot of phosphor knowledge, and so its unclear to me how the `Widget` getting attached/detached actually works.

**Comment by akosyakov** [comment]
If attachment/detachment is done properly via phorpshor APIs then everything should work. It basically flips the flag after attaching/detaching. I wonder why `isAttached` flag in the bogus state for you.
